pkg/router/template/router.go

```diff
@@ -530,16 +559,168 @@ func (r *templateRouter) DeleteServiceUnit(id string) {
+		if service, ok := r.findMatchingServiceUnit(key); ok {
+			newEndpoints := service.EndpointTable
+			glog.V(4).Infof("For new route backend %s, replacing endpoints for service %s: %+v", backendKey, key, service.EndpointTable)
+
+			weight, ok := newWeights[key]
+			if !ok {
+				weight = 0
+			}
+			if err := r.dynamicConfigManager.ReplaceRouteEndpoints(backendKey, oldEndpoints, newEndpoints, weight); err != nil {
+				glog.Warningf("Router will reload as there was an error dynamically replacing endpoints for new route backend %s, service %s: %v",
+					backendKey, key, err)
+				return false
+			}
+		}
+	}
+
+	glog.V(4).Infof("Dynamically added route backend %s", backendKey)
+	return true
+}
+
+// dynamicallyRemoveRoute attempts to dynamically remove a route.
+// Note: The config should have been synced at least once initially and
+//       the caller needs to acquire a lock [and release it].
+func (r *templateRouter) dynamicallyRemoveRoute(backendKey string, route *routeapi.Route, backend ServiceAliasConfig) bool {
+	if r.dynamicConfigManager == nil || !r.synced {
+		return false
+	}
+
+	glog.V(4).Infof("Dynamically removing route backend %s", backendKey)
+
+	if err := r.dynamicConfigManager.RemoveRoute(backendKey, route, backend.IsWildcard); err != nil {
+		glog.Warningf("Router will reload as there was an error dynamically removing a route backend %s: %v", backendKey, err)
+		return false
+	}
+
+	return true
+}
+
+// dynamicallyReplaceEndpoints attempts to dynamically replace endpoints
+// on all the routes associated with a given service.
+// Note: The config should have been synced at least once initially and
+//       the caller needs to acquire a lock [and release it].
+func (r *templateRouter) dynamicallyReplaceEndpoints(id string, service ServiceUnit, oldEndpoints []Endpoint) bool {
+	if r.dynamicConfigManager == nil || !r.synced {
+		return false
+	}
+
+	glog.V(4).Infof("Replacing endpoints dynamically for service %s", id)
+	newEndpoints := service.EndpointTable
+
+	// Update each of the routes that reference this service unit.
+	for backendKey := range service.ServiceAliasAssociations {
+		cfg, ok := r.state[backendKey]
+		if !ok {
+			glog.V(4).Infof("Associated service alias %s not found in state, ignoring ...", backendKey)
+			continue
+		}
+
+		// As the endpoints have changed, recalculate the weights.
+		newWeights := r.calculateServiceWeights(cfg.ServiceUnits)
+
+		// Get the weight for this service unit.
+		weight, ok := newWeights[id]
+		if !ok {
+			weight = 0
+		}
+
+		glog.V(4).Infof("Dynamically replacing endpoints for associated backend %s", backendKey)
+		if err := r.dynamicConfigManager.ReplaceRouteEndpoints(backendKey, oldEndpoints, newEndpoints, weight); err != nil {
+			// Error dynamically modifying the config, so return false to cause a reload to happen.
+			glog.Warningf("Router will reload as dynamic endpoint replacement for service id %s (backend=%s, weight=%v) failed: %v", id, backendKey, weight, err)
```

Comment:
You could easily spew errors here, i'm a little concerned with this in the long term.

Reply:
Will change to `glog.V(4).Infof`, so that its available if someone turns up the debugging level and for all the other "Router will reload" messages.